### PR TITLE
feat(vue-demo): add dynamic breadcrumbs

### LINKED ISSUE #139: [BUG] Category breadcrumbs path is missing
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Currently useBreadcrumbs() composable only returns names array for category breadcrumbs. Paths are missing so link to category cannot be created.

E.G.
```js
{
  "name": "Food"
}
```

Helper function to get breadcrumbs getCategoryBreadcrumbs() also gather only names:
```js
export function getCategoryBreadcrumbs(
  category: Category | null | undefined,
  options?: {
    /**
     * Start at specific index if your navigation
     * contains root names which should not be visible.
     */
    startIndex?: number;
  }
) {
  const breadcrumbs =
    category?.translated?.breadcrumb || category?.breadcrumb || [];
  const startIndex = options?.startIndex || 0;
  if (breadcrumbs.length <= startIndex) return [];
  return breadcrumbs.slice(startIndex).map((element) => {
    return {
      name: element,
    };
  });
}
```

Breadcrumbs in product.seoCategory / category API response also contain only names. So I think this is issue both in BE and FE parts:
Screenshot from 2023-04-14 00-07-12

### Expected Behavior

As declared in useBreadcrumbs return type I expect category breadcrumb with path returned:
```
type Breadcrumb = {
    name: string;
    path?: string;
};
```

### Steps To Reproduce

_No response_

### Environment

```markdown
- OS:
- Node:
- npm:
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by mdanilowicz**
### Description

This pull request adds a new request to fetch category breadcrumbs. The request is included on the category and product pages, while all other static pages build breadcrumbs with "local" data.

Added `getCategoryBreadcrumbs` method for fetching backend breadcrumbs in `useBreadcrumbs` composable
Added `buildDynamicBreadcrumbs` method for building breadcrumbs structure
Added `pushBreadcrumb` method to push single breadcrumb at the top of the breadcrumbs list
Added `associations` option to the `search` method in `useProductSearch` composable

closes #139

### CODE REVIEW

**Comment by BrocksiNet** [comment]
@kstala could you test this please :raised_hands:

File: templates/vue-demo-store/components/FrontendNavigationPage.vue [hunk 1]
```diff
@@ -30,11 +29,8 @@ if (!categoryResponse.value) {
   throw error.value;
 }
 
-const breadcrumbs = getCategoryBreadcrumbs(categoryResponse.value, {
-  startIndex: 2,
-});
-
-useBreadcrumbs(breadcrumbs);
+const { buildDynamicBreadcrumbs } = useBreadcrumbs();
+buildDynamicBreadcrumbs(props.navigationId);
```
Comment (patzick) [inline templates/vue-demo-store/components/FrontendNavigationPage.vue:33]: it's happening asynchronously, so I have two questions:
1. how are you assuring it's loaded on Server Side Render?
2. isn't there a blink on client side? What about showing previous breadcrumbs and replace them when new ones are loaded? Does it make sense?

File: packages/composables/src/useBreadcrumbs/useBreadcrumbs.ts [hunk 2]
```diff
@@ -48,8 +66,38 @@ export function useBreadcrumbs(
     _breadcrumbs.value = [];
   };
 
+  const getCategoryBreadcrumbs = async (categoryId: string) => {
+    const response = await apiClient.invoke(
+      "readBreadcrumb get /breadcrumb/{id}",
```
Comment (patzick) [inline packages/composables/src/useBreadcrumbs/useBreadcrumbs.ts:71]: this endpoint is not only to load category breadcrumbs, I'm not sure about exposing it here, what is the designed flow here looks like?

File: packages/composables/src/useBreadcrumbs/useBreadcrumbs.ts [hunk 3]
```diff
@@ -48,8 +66,38 @@ export function useBreadcrumbs(
...
+      {
+        pathParams: {
+          id: categoryId,
+        },
+      },
+    );
+    return response.data;
+  };
+
+  const buildDynamicBreadcrumbs = async (categoryId: string) => {
```
Comment (patzick) [inline packages/composables/src/useBreadcrumbs/useBreadcrumbs.ts:81]: what about product breadcrumbs?

File: packages/composables/src/useBreadcrumbs/useBreadcrumbs.ts [hunk 4]
```diff
@@ -48,8 +66,38 @@ export function useBreadcrumbs(
...
+      _breadcrumbs.value = breadcrumbs.breadcrumbs.map((breadcrumb) => {
+        // Adjust path to be compatible with the router
+        return {
+          ...breadcrumb,
+          path: `/${breadcrumb.path}`,
+        };
+      });
+    } catch (error) {
+      console.error("Error while fetching breadcrumbs", error);
+    }
```
Comment (patzick) [inline packages/composables/src/useBreadcrumbs/useBreadcrumbs.ts:94]: we should not handle errors in composable, it should be thrown to invoking component to handle in the project - analytics/notifications etc. however they need to handle it

File: .changeset/funny-hotels-watch.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+"@shopware-pwa/composables-next": minor
+---
+
+Added `getCategoryBreadcrumbs` method for fetching backend breadcrumbs in `useBreadcrumbs` composable
```
Comment (patzick) [inline .changeset/funny-hotels-watch.md:5]: the `buildDynamicBreadcrumbs` is more important functionality, should be mentioned and explained here more

File: packages/helpers/src/cms/getCmsBreadcrumbs.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
+/**
+ * Build the breadcrumbs for the CMS page
+ *
+ * @param page
+ * @returns
+ */
+export function getCmsBreadcrumbs<
```
Comment (patzick) [inline packages/helpers/src/cms/getCmsBreadcrumbs.ts:7]: not sure about this helper - it's not creating breadcrumbs, just displaying entity translated property, when would that be useful for the users?

**Review by patzick — COMMENTED** [review]
Good feature incoming! We need to clarify a few things here first :)

**Comment by patzick** [comment]
@mdanilowicz there are many unanswered comments in my previous CR. Please address them and explain the concept and the flow in the PR description, no need to rewrite it before agreeing on the API of these packages 🙌

File: packages/composables/src/useBreadcrumbs/useBreadcrumbs.ts (hunk 2 above)
Comment (patzick) [reply packages/composables/src/useBreadcrumbs/useBreadcrumbs.ts:79]: the endpoint is for both and takes more parameters
image


I just find `getCategoryBreadcrumbs` misleading as I would search for other types of breadcrumbs

Comment (patzick) [reply packages/composables/src/useBreadcrumbs/useBreadcrumbs.ts:79]: My point is - I would not keep this endpoint in the composable, it does not interfere with the logic and everyone can just use apiClient method for that and pass the data through `buildDynamicBreadcrumbs`

**Review by patzick — CHANGES_REQUESTED** [review]
I just have one more hesitation for useBreadcrumbs and I think the rest is in very good shape!

File: packages/composables/src/useBreadcrumbs/useBreadcrumbs.ts (hunk 2 above)
Comment (mdanilowicz) [reply packages/composables/src/useBreadcrumbs/useBreadcrumbs.ts]: done

**Review by BrocksiNet — APPROVED** [review]
There are some conflicts. But good work. :muscle:

File: .changeset/funny-hotels-watch.md [hunk 7]
```diff
@@ -0,0 +1,8 @@
...
```
Comment (patzick) [inline .changeset/funny-hotels-watch.md:5]:
```suggestion
```

**Review by patzick — APPROVED** [review]

**Pull request merged**